File: distributeCoinsInBinaryTree.java
// Time Complexity : O(n)
// Space Complexity :O(n)

class Solution {
    int moves;
    public int distributeCoins(TreeNode root) {
        if(root==null) return 0;
        dfs(root);
        return moves;
    }
    
    private int dfs(TreeNode root){
        //base
        if(root==null){
            return 0;
        }
        
        int extra= root.val-1;
        extra+=dfs(root.left);
        extra+=dfs(root.right);
        moves=moves+Math.abs(extra);
        
        return extra;
    }
}

File: distribute-coins-in-binary-tree.cpp
//Time - O(n)
//space - O(n)

class Solution {
public:
    int moves = 0;
    int distributeCoins(TreeNode* root) {
        if(root == NULL) return 0;
        dfs(root);
        return moves;
    }
    
    int dfs(TreeNode* root){
        if(root == NULL) return 0;
        
        int spareCoins = root->val - 1;
        spareCoins += dfs(root->left);
        spareCoins += dfs(root->right);
        moves += abs(spareCoins);
        return spareCoins;
    }
};